### ISSUE TITLE
`browser-launcher` is no longer maintained

### ISSUE THREAD

**Issue by seangenabe** [juliangruber/browser-run#112]
substack hasn't updated `browser-launcher` for 2 years; bugs aren't getting fixed. So I can't run this, for example, with Chrome on Windows.

I suggest using another package such as `browser-launcher2` instead.

**Comment by juliangruber** [comment]
yeah, i'm considering rewriting this logic once again, since browser-launcher2 also wasn't fully functioning for me. i was thinking about doing it @sindresorhus style, with one package per os and browser. i'll report back here!

**Comment by juliangruber** [comment]
- [ ] [osx chrome](https://github.com/juliangruber/osx-chrome)
- [ ] osx firefox
- [X] [osx electron](https://github.com/juliangruber/electron-stream)
- [X] [osx phantomjs](https://github.com/juliangruber/phantomjs-stream)
- [ ] osx safari
- [ ] linux chrome
- [ ] linux firefox
- [X] [linux electron](https://github.com/juliangruber/electron-stream)
- [X] [linux phantomjs](https://github.com/juliangruber/phantomjs-stream)
- [ ] windows chrome
- [ ] windows firefox
- [X] [windows electron](https://github.com/juliangruber/electron-stream)
- [X] [windows phantomjs](https://github.com/juliangruber/phantomjs-stream)
- [ ] windows edge

**Comment by juliangruber** [comment]
(btw i'm not looking forward to working in that windows vm, but i guess it just as to be done!)

**Comment by ashnur** [comment]
I have acces to windows 10 without running it in a vm, what has to be done in it?

**Comment by juliangruber** [comment]
development :P you could for example create `windows-chrome` which would be a windows port of `osx-chrome`

**Comment by ashnur** [comment]
what was the issue with `browser-launcher2`, is it related to browser detection? I am looking at https://github.com/vweevers/win-detect-browsers right now.

**Comment by juliangruber** [comment]
i have never used the module with windows, however i really want to have a modular solution to this problem, hence i'm creating the individual modules.

**Comment by ashnur** [comment]
i sympathize with the sentiment, however this will take longer :)

**Comment by juliangruber** [comment]
check out the todo list above :P

**Comment by ashnur** [comment]
i am not sure what's your point?

**Comment by juliangruber** [comment]
it's almost done already

**Comment by ashnur** [comment]
ok, so how should I detect google chrome on windows?

**Comment by juliangruber** [comment]
no need to detect it, just figure out the path and launch it with the right parameters. check out browser-launcher for inspiration

**Comment by ashnur** [comment]
Are you referring to this https://github.com/substack/browser-launcher/blob/master/lib/windows/chrome.js? Because https://gist.github.com/vweevers/e8c3863759fd8f9659ce

**Comment by juliangruber** [comment]
yeah, just do what works!

**Comment by juliangruber** [comment]
@ashnur linux work is complete for now. how far are you with windows?

**Comment by ashnur** [comment]
Should be able to finish windows-chrome later today, after work :).

**Comment by ashnur** [comment]
Ok, so apart from me promising stuff irresponsibly (stuff came up :( ), I am almost done, but I would like to test it. And then I would like to know why @juliangruber didn't use the `cb` argument in `linux-chrome`, and if it's ok if I will use it in the windows modules because I understand that I don't have to detect, "just figure out the path", however that's also nowhere trivial on windows. Basically it's a bunch of guesswork, and I will have to use an async interface. I am open to suggestions if anything I said is incorrect or if you see a better way.

**Comment by juliangruber** [comment]
linux-chrome doesn't have a `cb` because it doesn't do any asynchronous work, it can just return the browser process object right away. If you need to do async work, use a callback :)

**Comment by ashnur** [comment]
Cool, thanks for the clarification. Will do that then. Btw, does anyone here have a suggestion how I could write a test for this? Because then I could perhaps run a few automated tests on different version of windows not just Win 10.

**Comment by juliangruber** [comment]
create a nodejs http server, launch a browser pointing at its address, see if the request comes in

**Comment by ashnur** [comment]
ok, the test passes now, and the browser starts seemingly properly but I am still very unhappy about the whole situation. I want to do some further testing and removing the bloat. https://github.com/ashnur/windows-chrome 
I am fairly sure there are several things that are still completely wrong about what I did but at least I know it's going to work :)

**Comment by juliangruber** [comment]
cool, i'll start integrating it

**Comment by juliangruber** [comment]
...once you publish it :)

**Comment by ashnur** [comment]
https://www.npmjs.com/package/windows-chrome

**Comment by eugeneware** [comment]
@juliangruber I've written the Windows Edge module to keep this moving. Tested in Windows 10 in a VM.

https://github.com/eugeneware/windows-edge

**Comment by juliangruber** [comment]
only one browser left! windows-firefox 🎆

**Comment by matthiasg** [comment]
can this mode already be enabled through some option ? right now it still defaults to browser-launcher. 

PS: Related, what about chrome headless mode ?

**Comment by refactorized** [comment]
I would really like to see this especially if it enables better headless config via headless chrome for instance, ( which is really just a start-up param ).  What about falling back to browser-launch for firefox/win in the mean time?  I would look at taking on the firefox/win issue, but have no stable windows env to work on atm.  

Maybe one of those free MS VMs hmmm.

**Comment by vweevers** [comment]
I'll create `windows-firefox`.

**Comment by vweevers** [comment]
Published: [`windows-firefox`](https://www.npmjs.com/package/windows-firefox). Supports headless mode and filtering installed versions by release channel or version. Also made [`create-firefox-profile`](https://www.npmjs.com/package/create-firefox-profile).
